### PR TITLE
Fallback to plain INSERT

### LINKED ISSUE #118: Make JDBC table manager configurable to support drivers without prepared statement
It would be useful to have a table manager for Presto that does not depend on HDFS and Hive.  This will require loading data via individual INSERT statements.  Obviously performance will be poor, but for very small tables it should be acceptable.

### PR DESCRIPTION
**Pull request by kokosing**
Fallback to plain INSERT

When JDBC driver does not support PreparedStatement JdbcTableManager
will fallback to insert each row with separate statement.

Fixes #118

### CODE REVIEW

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java [hunk 1]
```diff
@@ -93,9 +94,11 @@ public TableInstance createImmutable(JdbcTableDefinition tableDefinition, TableH
             // could be a table from another schema that has the same name
             dropTableIgnoreError(tableName);
             createAndInsertData(tableDefinition, tableName);
-        } else if (!tableExists(tableName)) {
+        }
```
Comment (maciejgrzybek) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java]: Unrelated change.

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java [hunk 2]
```diff
@@ -199,58 +202,171 @@ public String getDatabaseName()
...
-        for (int rowIndex = 0; rowIndex < insertCounts.length; ++rowIndex) {
-            if (insertCounts[rowIndex] != 1) {
-                throw new RuntimeException("could not insert values for row " + (baseBatchRowIndex + rowIndex) + "; values=" + batchRows.get(rowIndex));
+
+        @Override
+        public void load(List<List<Object>> batch)
+                throws SQLException
+        {
+            for (List<Object> row : batch) {
+                queryExecutor.executeQuery(insertSql, asQueryParams(row).toArray(new QueryExecutor.QueryParam[0]));
```
Comment (maciejgrzybek) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java]: Wouldn't `QueryExecutor` try using prepared statements when `QueryParams` are given?
`JdbcQueryExecutor` seems base on `PreparedStatement` in `executeQueryWithParams` method.

Why don't use just values without `QueryParams`? It would also save us issuing additional query (here: https://github.com/prestodb/tempto/pull/150/commits/8a8cc07040617168e94f94bffac7d407aac8bba8#diff-33d694596851b3e9776aefc8165805ffR338).

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java [hunk 3]
```diff
@@ -199,58 +202,171 @@ public String getDatabaseName()
...
+        }
+        catch (SQLException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private void insertData(TableName tableName, Iterator<List<Object>> dataRows)
+            throws SQLException
+    {
+        Loader loader = null;
```
Comment (maciejgrzybek) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java]: Move `columnsCount` before loop and use `try-with-resources` since `Loader` is auto-closeable, maybe?

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java [hunk 4]
```diff
@@ -199,58 +202,171 @@ public String getDatabaseName()
...
+            public boolean hasNext()
+            {
+                return dataRows.hasNext();
+            }
+
+            @Override
+            public List<List<Object>> next()
+            {
+                List<List<Object>> batch = new ArrayList<>();
+                while (dataRows.hasNext() || batch.size() < BATCH_SIZE) {
```
Comment (maciejgrzybek) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java]: If `dataRows` is empty (`hasNext() = false`) but `batch.size()` < `BATCH_SIZE`, there will be `NoSuchElementException` thrown in next line.

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java [hunk 5]
```diff
@@ -199,58 +202,171 @@ public String getDatabaseName()
...
+                LOGGER.debug("Unable to insert data with PreparedStatement", sqlException);
+                return new InsertLoader(queryExecutor, insertSql, tableName);
+            }
+        }
+    }
+
+    private static class BatchLoader
+            implements Loader
+    {
+        private PreparedStatement preparedStatement;
```
Comment (maciejgrzybek) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java]: `final`

File: tempto-examples/src/main/resources/sql-tests/datasets/sample_jdbc_table.ddl [hunk 6]
```diff
@@ -1,5 +1,5 @@
 -- type: jdbc
 CREATE TABLE %NAME% (
-  id INT,
+  id BIGINT,
```
Comment (maciejgrzybek) [inline tempto-examples/src/main/resources/sql-tests/datasets/sample_jdbc_table.ddl]: Unrelated change.

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java (hunk 3 above)
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java]: that was my initial idea, but to create `Loader` I need to read at least one `batch`. I need to do some refactor to make it possible

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java (hunk 4 above)
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java]: good catch! thanks

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java (hunk 2 above)
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/JdbcTableManager.java]: Yeah... I didn't notice that `PreparedStatement` is used there.

Did you mean to create a sql query string by concatenation of values?  This is not going to work, I would need to handle each type (int, floats, string)

File: tempto-examples/src/main/resources/sql-tests/datasets/sample_jdbc_table.ddl (hunk 6 above)
Comment (kokosing) [inline tempto-examples/src/main/resources/sql-tests/datasets/sample_jdbc_table.ddl]: it is related I used this table to insert data for with presto jdbc table manager

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/BatchLoader.java [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+    }
+
+    @Override
+    public void load(List<List<Object>> batch)
+            throws SQLException
+    {
+        for (List<Object> row : batch) {
+            int pos = 1;
+            for (Object value : row) {
+                preparedStatement.setObject(pos, value);
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/BatchLoader.java]: checkArgument that number of objects in row matches columnsCount

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/BatchLoader.java [hunk 8]
```diff
@@ -0,0 +1,78 @@
...
+            }
+            preparedStatement.addBatch();
+        }
+        executeBatchWithVerification(preparedStatement, batch);
+    }
+
+    private void executeBatchWithVerification(PreparedStatement preparedStatement, List<List<Object>> batch)
+            throws SQLException
+    {
+        if (batch.size() == 0) {
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/BatchLoader.java]: move this condition to load

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/BatchLoader.java [hunk 9]
```diff
@@ -0,0 +1,78 @@
...
+            for (Object value : row) {
+                preparedStatement.setObject(pos, value);
+                ++pos;
+            }
+            preparedStatement.addBatch();
+        }
+        executeBatchWithVerification(preparedStatement, batch);
+    }
+
+    private void executeBatchWithVerification(PreparedStatement preparedStatement, List<List<Object>> batch)
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/BatchLoader.java]: maybe inline this method to load. The split is weird and arbitrary.

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/LoaderFactory.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+    Loader create(QueryExecutor queryExecutor, String tableName)
+            throws SQLException
+    {
+        List<JDBCType> columnTypes = queryExecutor.executeQuery("SELECT * FROM " + tableName + " LIMIT 1").getColumnTypes();
+
+        try {
+            return new BatchLoader(queryExecutor, tableName, columnTypes.size());
+        }
+        catch (SQLException sqlException) {
+            LOGGER.debug("Unable to insert data with PreparedStatement", sqlException);
```
Comment (losipiuk) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/LoaderFactory.java]: LOGGER.warn

File: tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/BatchLoader.java (hunk 7 above)
Comment (kokosing) [inline tempto-core/src/main/java/com/teradata/tempto/internal/fulfillment/table/jdbc/BatchLoader.java]: done

File: tempto-examples/src/main/resources/sql-tests/datasets/sample_jdbc_table.ddl (hunk 6 above)
Comment (maciejgrzybek) [inline tempto-examples/src/main/resources/sql-tests/datasets/sample_jdbc_table.ddl]: So why INT didn't work here?

Comment (kokosing) [inline tempto-examples/src/main/resources/sql-tests/datasets/sample_jdbc_table.ddl]: This table definition was used for psql tables and tests were returning INT. While INT in Presto is not yet supported, at least it was not in 0.143 which was used for tempto testing.

Comment (maciejgrzybek) [inline tempto-examples/src/main/resources/sql-tests/datasets/sample_jdbc_table.ddl]: That's why I said it's unrelated change. ;) It could be a separate commit. Though, it's such a small change it's not worth it.
